### PR TITLE
[ISSUE-3128] Support Projection push down for MemoryTable Engine

### LINKED ISSUE #3128: Engine Memory support projection
**Summary**

ontime_memory used `engine = Memory` cost extra read on used column, it can support column projection like FuseEngine.
```
 

:) select count(1) ,avg(Year), sum(DayOfWeek)  from ontime_memory
    -> ;
+----------+-----------+----------------+
| count(1) | avg(Year) | sum(DayOfWeek) |
+----------+-----------+----------------+
|  3342223 |      2020 |       13114669 |
+----------+-----------+----------------+
1 row in set (0.04 sec)
Read 3342223 rows, 2.44 GB in 0.033 sec., 102.81 million rows/sec., 75.05 GB/sec

:) select count(1) ,avg(Year), sum(DayOfWeek)  from ontime_fuse;
+----------+-----------+----------------+
| count(1) | avg(Year) | sum(DayOfWeek) |
+----------+-----------+----------------+
|  2599509 |      2020 |       10200307 |
+----------+-----------+----------------+
1 row in set (1.80 sec)
Read 2599509 rows, 7.8 MB in 1.787 sec., 1.46 million rows/sec., 4.37 MB/sec.


:) explain select count(1) ,avg(Year), sum(DayOfWeek)  from ontime_fuse;
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                                                                               |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection: count(1):UInt64, avg(Year):Float64, sum(DayOfWeek):UInt64                                                                                                                 |
|   AggregatorFinal: groupBy=[[]], aggr=[[count(1), avg(Year), sum(DayOfWeek)]]                                                                                                         |
|     AggregatorPartial: groupBy=[[]], aggr=[[count(1), avg(Year), sum(DayOfWeek)]]                                                                                                     |
|       Expression: 1:UInt8, Year:UInt16, DayOfWeek:UInt8 (Before GroupBy)                                                                                                              |
|         ReadDataSource: scan partitions: [260], scan schema: [Year:UInt16, DayOfWeek:UInt8], statistics: [read_rows: 2599509, read_bytes: 7798527], push_downs: [projections: [0, 4]] |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+


```

### PR DESCRIPTION
**Pull request by ygf11**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

MemoryTable supports projection push down in select and explain select statement.

## Changelog

- New Feature

## Related Issues

Fixes #3128 

## Test Plan

Unit Tests

### CODE REVIEW

File: query/src/storages/memory/memory_table_test.rs [hunk 1]
```diff
@@ -63,11 +63,42 @@ async fn test_memorytable() -> Result<()> {
             .unwrap();
     }
 
-    // read.
+    // read pruned column.
+    {
+        let mut push_downs = Extras::default();
+        push_downs.projection = Some(vec![0usize]);
```
Comment (zhang2014) [inline query/src/storages/memory/memory_table_test.rs:69]: Maybe we should also test `projection 1`?

File: query/src/storages/memory/memory_table.rs [hunk 2]
```diff
@@ -93,13 +122,33 @@ impl Table for MemoryTable {
...
+        plan: &ReadDataSourcePlan,
     ) -> Result<SendableDataBlockStream> {
-        let blocks = self.blocks.read();
-        Ok(Box::pin(MemoryTableStream::try_create(
-            ctx,
-            blocks.clone(),
-        )?))
+        let push_downs = &plan.push_downs;
+
+        let blocks = if let Some(Extras {
```
Comment (zhang2014) [inline query/src/storages/memory/memory_table.rs:129]: maybe use match is better?

File: query/src/storages/memory/memory_table.rs [hunk 3]
```diff
@@ -71,17 +73,44 @@ impl Table for MemoryTable {
...
         &self,
         ctx: Arc<QueryContext>,
-        _push_downs: Option<Extras>,
+        push_downs: Option<Extras>,
     ) -> Result<(Statistics, Partitions)> {
         let blocks = self.blocks.read();
 
-        let rows = blocks.iter().map(|block| block.num_rows()).sum();
-        let bytes = blocks.iter().map(|block| block.memory_size()).sum();
+        let statistics = if let Some(Extras {
```
Comment (zhang2014) [inline query/src/storages/memory/memory_table.rs:87]: maybe use match is better?

Comment (ygf11) [reply query/src/storages/memory/memory_table.rs:87]: Yes, match is better for future work. let me fix it.

File: query/src/storages/memory/memory_table_test.rs [hunk 4]
```diff
@@ -63,15 +63,41 @@ async fn test_memorytable() -> Result<()> {
...
+    // read tests
     {
-        let source_plan = table.read_plan(ctx.clone(), None).await?;
-        ctx.try_set_partitions(source_plan.parts.clone())?;
-        assert_eq!(table.engine(), "Memory");
-
-        let stream = table.read(ctx.clone(), &source_plan).await?;
-        let result = stream.try_collect::<Vec<_>>().await?;
-        assert_blocks_sorted_eq(
+        let push_downs_vec: Vec<Option<Extras>> =
```
Comment (zhang2014) [inline query/src/storages/memory/memory_table_test.rs:68]: 👍

**Review by zhang2014 — APPROVED** [review]

**Comment by ygf11** [comment]
@zhang2014 @BohuTANG Thanks for your help:)

**Pull request merged**